## **RESOLUCIÓN EVALUACIÓN MÓDULO 2:**
#### **ANÁLISIS DE DATOS**

- Librerías

In [1]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Libreria Regex para buscar patrones
import re

# Imputación de nulos usando métodos avanzados estadísticos
# -----------------------------------------------------------------------
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.impute import KNNImputer

# Librerías de visualización
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt
# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

#### **Fase 1: Exploración y Limpieza**

**1. Exploración Inicial:**

• Realiza una exploración inicial de los datos para identificar posibles
problemas, como valores nulos, atípicos o datos faltantes en las columnas
relevantes.

• Utiliza funciones de Pandas para obtener información sobre la estructura
de los datos, la presencia de valores nulos y estadísEcas básicas de las
columnas involucradas.

• Une los dos conjuntos de datos de la forma más eficiente.

- **Lectura archivos csv**

In [2]:
df_flight = pd.read_csv("Customer Flight Activity.csv", index_col = None)
df_flight.head()

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0          100018  2017      1               3                        0   
1          100102  2017      1              10                        4   
2          100140  2017      1               6                        0   
3          100214  2017      1               0                        0   
4          100272  2017      1               0                        0   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1             14      2030               203.0                0   
2              6      1200               120.0                0   
3              0         0                 0.0                0   
4              0         0                 0.0                0   

   Dollar Cost Points Redeemed  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0

In [3]:
df_loyalty = pd.read_csv("Customer Loyalty History.csv", index_col = None)
df_loyalty.head()

Loyalty Number Country          Province       City Postal Code  Gender  \
0          480934  Canada           Ontario    Toronto     M2Z 4K1  Female   
1          549612  Canada           Alberta   Edmonton     T3G 6Y6    Male   
2          429460  Canada  British Columbia  Vancouver     V6E 3D9    Male   
3          608370  Canada           Ontario    Toronto     P1W 1K4    Male   
4          530508  Canada            Quebec       Hull     J8Y 3Z5    Male   

  Education    Salary Marital Status Loyalty Card      CLV Enrollment Type  \
0  Bachelor   83236.0        Married         Star  3839.14        Standard   
1   College       NaN       Divorced         Star  3839.61        Standard   
2   College       NaN         Single         Star  3839.75        Standard   
3   College       NaN         Single         Star  3839.75        Standard   
4  Bachelor  103495.0        Married         Star  3842.79        Standard   

   Enrollment Year  Enrollment Month  Cancellation Year  Cancellation Month  
0             2016                 2                NaN                 NaN  
1             2016                 3                NaN                 NaN  
2             2014                 7             2018.0                 1.0  
3             2013                 2                NaN                 NaN  
4             2014                10                NaN                 NaN

- Creación copias de los Dataframes para trabajar con seguridad

In [4]:
df_flight_copy = df_flight.copy()
df_flight_copy.head(2)

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0          100018  2017      1               3                        0   
1          100102  2017      1              10                        4   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1             14      2030               203.0                0   

   Dollar Cost Points Redeemed  
0                            0  
1                            0

In [5]:
df_loyalty_copy = df_loyalty.copy()
df_loyalty_copy.head(2)

Loyalty Number Country Province      City Postal Code  Gender Education  \
0          480934  Canada  Ontario   Toronto     M2Z 4K1  Female  Bachelor   
1          549612  Canada  Alberta  Edmonton     T3G 6Y6    Male   College   

    Salary Marital Status Loyalty Card      CLV Enrollment Type  \
0  83236.0        Married         Star  3839.14        Standard   
1      NaN       Divorced         Star  3839.61        Standard   

   Enrollment Year  Enrollment Month  Cancellation Year  Cancellation Month  
0             2016                 2                NaN                 NaN  
1             2016                 3                NaN                 NaN

- **Unión de Dataframes en uno sólo**

Se utiliza el método ".merge()" de tipo "inner" ya que como los dos Dataframes comparten una columna con el mismo nombre "Loyalty Number" siendo el mismo tipo de dato y haciendo referencia al mismo signifivado; la unión de dichos Dataframes se hará a través de ésta.

In [6]:
df_aerolinea = pd.merge(df_flight_copy, df_loyalty_copy, on="Loyalty Number")

- Comprobación de la unión de los Dataframes en uno sólo.

In [7]:
df_aerolinea.head()

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0          100018  2017      1               3                        0   
1          100018  2017      2               2                        2   
2          100018  2018     10               6                        4   
3          100018  2017      4               4                        0   
4          100018  2017      5               0                        0   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1              4      1320               132.0                0   
2             10      3110               311.0              385   
3              4       924                92.0                0   
4              0         0                 0.0                0   

   Dollar Cost Points Redeemed Country Province      City Postal Code  Gender  \
0                            0  Canada  Alberta  Edmonton     T9G 1W3  Female   
1                            0  Canada  Alberta  Edmonton     T9G 1W3  Female   
2                           31  Canada  Alberta  Edmonton     T9G 1W3  Female   
3                            0  Canada  Alberta  Edmonton     T9G 1W3  Female   
4                            0  Canada  Alberta  Edmonton     T9G 1W3  Female   

  Education   Salary Marital Status Loyalty Card     CLV Enrollment Type  \
0  Bachelor  92552.0        Married       Aurora  7919.2        Standard   
1  Bachelor  92552.0        Married       Aurora  7919.2        Standard   
2  Bachelor  92552.0        Married       Aurora  7919.2        Standard   
3  Bachelor  92552.0        Married       Aurora  7919.2        Standard   
4  Bachelor  92552.0        Married       Aurora  7919.2        Standard   

   Enrollment Year  Enrollment Month  Cancellation Year  Cancellation Month  
0             2016                 8                NaN                 NaN  
1             2016                 8                NaN                 NaN  
2             2016                 8                NaN                 NaN  
3             2016                 8                NaN                 NaN  
4             2016                 8                NaN                 NaN

- Creación copia de nuevo Dataframe para trabajar con seguridad

In [10]:
df_aerolinea_copy = df_aerolinea.copy()
df_aerolinea_copy.sample(10)

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
379729          188307  2017      3              16                        4   
188792          559879  2017      9               2                        0   
173708          520137  2018      8               0                        0   
305915          853704  2017     10               6                        2   
4868            111788  2018      8              13                        2   
82176           298853  2017      1               8                        2   
292800          820223  2017      1               0                        0   
184812          549858  2018      2               0                        0   
404655          974441  2018      4               0                        0   
36207           187887  2018      2               5                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
379729             20      2540               254.0                0   
188792              2      3320               332.0                0   
173708              0         0                 0.0                0   
305915              8      3040               304.0                0   
4868               15      1395               139.0              494   
82176              10      2880               288.0                0   
292800              0         0                 0.0                0   
184812              0         0                 0.0                0   
404655              0         0                 0.0                0   
36207               5      2270               340.5                0   

        Dollar Cost Points Redeemed Country          Province            City  \
379729                            0  Canada           Ontario         Toronto   
188792                            0  Canada           Ontario         Toronto   
173708                            0  Canada  British Columbia  West Vancouver   
305915                            0  Canada  British Columbia        Victoria   
4868                             40  Canada  British Columbia       Vancouver   
82176                             0  Canada           Ontario         Toronto   
292800                            0  Canada  British Columbia    Dawson Creek   
184812                            0  Canada           Ontario         Toronto   
404655                            0  Canada           Ontario          Ottawa   
36207                             0  Canada            Quebec     Quebec City   

       Postal Code  Gender             Education    Salary Marital Status  \
379729     P1J 8T7  Female               College       NaN        Married   
188792     M2M 7K8  Female              Bachelor   59382.0        Married   
173708     V6V 8Z3    Male  High School or Below   54162.0        Married   
305915     V10 6T5    Male              Bachelor   50920.0        Married   
4868       V6E 3Z3  Female              Bachelor   53537.0        Married   
82176      M8Y 4K8  Female              Bachelor  102403.0        Married   
292800     U5I 4F1    Male                Doctor  248809.0        Married   
184812     M8Y 4K8  Female              Bachelor  100159.0        Married   
404655     K1F 2R2  Female               College       NaN         Single   
36207      G1B 3L5    Male              Bachelor   90480.0        Married   

       Loyalty Card       CLV Enrollment Type  Enrollment Year  \
379729         Nova   3164.49        Standard             2014   
188792         Nova   6457.56        Standard             2013   
173708         Nova   3716.48        Standard             2015   
305915       Aurora  10721.61        Standard             2016   
4868           Nova   4935.19        Standard             2016   
82176          Nova   2746.12        Standard             2014   
292800         Star   4145.71        Standard             2017   
184812         Star   3861.49        Standard             2012   


- **Análisis Exploratorio de los Datos del Dataframe**

In [13]:
# Nombre delas columnas 

df_aerolinea_copy.columns

Index(['Loyalty Number', 'Year', 'Month', 'Flights Booked',
       'Flights with Companions', 'Total Flights', 'Distance',
       'Points Accumulated', 'Points Redeemed', 'Dollar Cost Points Redeemed',
       'Country', 'Province', 'City', 'Postal Code', 'Gender', 'Education',
       'Salary', 'Marital Status', 'Loyalty Card', 'CLV', 'Enrollment Type',
       'Enrollment Year', 'Enrollment Month', 'Cancellation Year',
       'Cancellation Month'],
      dtype='object')

In [15]:
# Número de filas/ registros

df_aerolinea_copy.shape[0]

405624

In [16]:
# Número de columnas

df_aerolinea_copy.shape[1]

25

In [ ]:
# Información general del Datafarme:número de registros, valores nulos/ no nulos, tipo de dato

df_aerolinea_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 405624 entries, 0 to 405623
Data columns (total 25 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
 10  Country                      405624 non-null  object 
 11  Province                     405624 non-null  object 
 12  City                         405624 non-null  object 
 13 

In [14]:
df_aerolinea_copy.isnull().sum()

Loyalty Number                      0
Year                                0
Month                               0
Flights Booked                      0
Flights with Companions             0
Total Flights                       0
Distance                            0
Points Accumulated                  0
Points Redeemed                     0
Dollar Cost Points Redeemed         0
Country                             0
Province                            0
City                                0
Postal Code                         0
Gender                              0
Education                           0
Salary                         102672
Marital Status                      0
Loyalty Card                        0
CLV                                 0
Enrollment Type                     0
Enrollment Year                     0
Enrollment Month                    0
Cancellation Year              355560
Cancellation Month             355560
dtype: int64

In [17]:
# Filas/ registros duplicados

df_aerolinea_copy.duplicated().sum()

1864

In [21]:
# Columnas duplicadas

df_aerolinea_copy.columns.duplicated().sum()

0

In [35]:
# Columnas categóricas 

col_categoricas = df_aerolinea_copy.select_dtypes(include = "O").columns
col_categoricas 

Index(['Country', 'Province', 'City', 'Postal Code', 'Gender', 'Education',
       'Marital Status', 'Loyalty Card', 'Enrollment Type'],
      dtype='object')

In [33]:
# Columnas categóricas con valores nulos

col_cat_nulos = df_aerolinea_copy[df_aerolinea_copy.columns[df_aerolinea_copy.isnull().any()]].select_dtypes(include = "O").columns
col_cat_nulos

Index([], dtype='object')

In [29]:
# Columnas numéricas con valores nulos

col_num_nulos = df_aerolinea_copy[df_aerolinea_copy.columns[df_aerolinea_copy.isnull().any()]].select_dtypes(include = np.number).columns

col_num_nulos

Index(['Salary', 'Cancellation Year', 'Cancellation Month'], dtype='object')

In [36]:
# Columnas numéricas tipo "int"

col_num_int = df_aerolinea_copy.select_dtypes(include = "int").columns
col_num_int 

Index(['Loyalty Number', 'Year', 'Month', 'Flights Booked',
       'Flights with Companions', 'Total Flights', 'Distance',
       'Points Redeemed', 'Dollar Cost Points Redeemed', 'Enrollment Year',
       'Enrollment Month'],
      dtype='object')

In [30]:
# Columnas numéricas tipo "int" con valores nulos

col_int_nulos = df_aerolinea_copy[df_aerolinea_copy.columns[df_aerolinea_copy.isnull().any()]].select_dtypes(include = "int").columns
col_int_nulos

Index([], dtype='object')

In [37]:
# Columnas numéricas tipo "int"

col_num_float = df_aerolinea_copy.select_dtypes(include = "float").columns
col_num_float

Index(['Points Accumulated', 'Salary', 'CLV', 'Cancellation Year',
       'Cancellation Month'],
      dtype='object')

In [31]:
# Columnas numéricas tipo "float" con valores nulos:

col_float_nulos = df_aerolinea_copy[df_aerolinea_copy.columns[df_aerolinea_copy.isnull().any()]].select_dtypes(include = "float").columns
col_float_nulos

Index(['Salary', 'Cancellation Year', 'Cancellation Month'], dtype='object')

In [20]:
# Principales estadísticos

df_aerolinea_copy.describe()

Loyalty Number           Year          Month  Flights Booked  \
count   405624.000000  405624.000000  405624.000000   405624.000000   
mean    550037.873084    2017.500000       6.500000        4.115052   
std     258935.286969       0.500001       3.452057        5.225518   
min     100018.000000    2017.000000       1.000000        0.000000   
25%     326961.000000    2017.000000       3.750000        0.000000   
50%     550834.000000    2017.500000       6.500000        1.000000   
75%     772194.000000    2018.000000       9.250000        8.000000   
max     999986.000000    2018.000000      12.000000       21.000000   

       Flights with Companions  Total Flights       Distance  \
count            405624.000000  405624.000000  405624.000000   
mean                  1.031805       5.146858    1208.880059   
std                   2.076869       6.521227    1433.155320   
min                   0.000000       0.000000       0.000000   
25%                   0.000000       0.000000       0.000000   
50%                   0.000000       1.000000     488.000000   
75%                   1.000000      10.000000    2336.000000   
max                  11.000000      32.000000    6293.000000   

       Points Accumulated  Points Redeemed  Dollar Cost Points Redeemed  \
count       405624.000000    405624.000000                405624.000000   
mean           123.692721        30.696872                     2.484503   
std            146.599831       125.486049                    10.150038   
min              0.000000         0.000000                     0.000000   
25%              0.000000         0.000000                     0.000000   
50%             50.000000         0.000000                     0.000000   
75%            239.000000         0.000000                     0.000000   
max            676.500000       876.000000                    71.000000   

              Salary            CLV  Enrollment Year  Enrollment Month  \
count  302952.000000  405624.000000    405624.000000     405624.000000   
mean    79268.825953    7991.976226      2015.252529          6.668008   
std     34992.133508    6863.663857         1.979427          3.399766   
min    -58486.000000    1898.010000      2012.000000          1.000000   
25%     59262.000000    3985.320000      2014.000000          4.000000   
50%     73479.000000    5776.340000      2015.000000          7.000000   
75%     88612.000000    8936.820000      2017.000000         10.000000   
max    407228.000000   83325.380000      2018.000000         12.000000   

       Cancellation Year  Cancellation Month  
count       50064.000000        50064.000000  
mean         2016.499521            6.966443  
std             1.384336            3.448300  
min          2013.000000            1.000000  
25%          2016.000000            4.000000  
50%          2017.000000            7.000000  
75%          2018.000000           10.000000  
max          2018.000000           12.000000

In [22]:
# Principales estadísticos columas categóricas ("object")

df_aerolinea_copy.describe(include = "object")

Country Province     City Postal Code  Gender Education Marital Status  \
count   405624   405624   405624      405624  405624    405624         405624   
unique       1       11       29          55       2         5              3   
top     Canada  Ontario  Toronto     V6E 3D9  Female  Bachelor        Married   
freq    405624   130896    81096       21984  203640    253752         235800   

       Loyalty Card Enrollment Type  
count        405624          405624  
unique            3               2  
top            Star        Standard  
freq         184224          382200

In [23]:
# Principales estadísticos columas numéricas de tipo "int"

df_aerolinea_copy.describe(include = "int64")

Loyalty Number           Year          Month  Flights Booked  \
count   405624.000000  405624.000000  405624.000000   405624.000000   
mean    550037.873084    2017.500000       6.500000        4.115052   
std     258935.286969       0.500001       3.452057        5.225518   
min     100018.000000    2017.000000       1.000000        0.000000   
25%     326961.000000    2017.000000       3.750000        0.000000   
50%     550834.000000    2017.500000       6.500000        1.000000   
75%     772194.000000    2018.000000       9.250000        8.000000   
max     999986.000000    2018.000000      12.000000       21.000000   

       Flights with Companions  Total Flights       Distance  Points Redeemed  \
count            405624.000000  405624.000000  405624.000000    405624.000000   
mean                  1.031805       5.146858    1208.880059        30.696872   
std                   2.076869       6.521227    1433.155320       125.486049   
min                   0.000000       0.000000       0.000000         0.000000   
25%                   0.000000       0.000000       0.000000         0.000000   
50%                   0.000000       1.000000     488.000000         0.000000   
75%                   1.000000      10.000000    2336.000000         0.000000   
max                  11.000000      32.000000    6293.000000       876.000000   

       Dollar Cost Points Redeemed  Enrollment Year  Enrollment Month  
count                405624.000000    405624.000000     405624.000000  
mean                      2.484503      2015.252529          6.668008  
std                      10.150038         1.979427          3.399766  
min                       0.000000      2012.000000          1.000000  
25%                       0.000000      2014.000000          4.000000  
50%                       0.000000      2015.000000          7.000000  
75%                       0.000000      2017.000000         10.000000  
max                      71.000000      2018.000000         12.000000

In [24]:
# Principales estadísticos columas numéricas de tipo "float"

df_aerolinea_copy.describe(include = "float")

Points Accumulated         Salary            CLV  Cancellation Year  \
count       405624.000000  302952.000000  405624.000000       50064.000000   
mean           123.692721   79268.825953    7991.976226        2016.499521   
std            146.599831   34992.133508    6863.663857           1.384336   
min              0.000000  -58486.000000    1898.010000        2013.000000   
25%              0.000000   59262.000000    3985.320000        2016.000000   
50%             50.000000   73479.000000    5776.340000        2017.000000   
75%            239.000000   88612.000000    8936.820000        2018.000000   
max            676.500000  407228.000000   83325.380000        2018.000000   

       Cancellation Month  
count        50064.000000  
mean             6.966443  
std              3.448300  
min              1.000000  
25%              4.000000  
50%              7.000000  
75%             10.000000  
max             12.000000

**OBSERVACIONES SIGNIFICATIVAS:**

- **Número de columnas:** 25 --> No hay duplicados.
- **Número de filas/ registros:** 405624 --> 1864 duplicados.
- **Columas categóricas:** "Country", "Province", "City", "Postal Code", "Gender", "Education", "Marital Status", "Loyalty Card", "Enrollment Type".
- Columnas categóricas con valores nulos: No hay.
- **Columnas numéricas tipo "int":** "Loyalty Number", "Year", "Month", "Flights" "Booked", "Flights with Companions", "Total Flights", "Distance",          "Points Redeemed", "Dollar Cost Points Redeemed", "Enrollment Year",    "Enrollment Month".
- Columnas numéricas tipo "int" con valores nulos: No hay.
- **Columnas numéricas tipo "float":** "Points Accumulated", "Salary", "CLV", "Cancellation Year", "Cancellation Month".
- Columnas numéricas tipo "float" con valores nulos: 
    "Salary", "Cancellation Year", "Cancellation Month".